### ISSUE TITLE
Visual Studio crashes when working with large files.

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#48592]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/814366/visual-studio-crashes-when-working-with-large-file.html)._

---
I have a class with a single dictionary in a class that generates some static data. It's about 32000 lines :) I am able to edit it in Visual Studio "Code" and its very responsive but every time I try to edit it in Visual Studio,  it brings it to a halt then crashes. I want to be able to edit it in visual studio to keep with my workflow instead of switching context everytime.

---
#### Feedback Bot on 11/11/2019, 00:47 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Sam Harwell [MSFT] on 11/12/2019, 02:19 PM: 

<p>Thank you for providing feedback, and we’re sorry to hear it’s not behaving as you expect. Based on your description so far, it sounds like you are experiencing a problem which is historically hard to diagnose and resolve using the normal “steps to reproduce”. We created a set of instructions for providing additional information which will help us track down the true source of the problems.</p>
<p>Based on the information provided so far, the most likely scenario to follow is for “Crashes”. Please take a look at the following document to provide the feedback most relevant for the problems you would like to see fixed: <a target='_blank' href="https://github.com/dotnet/roslyn/wiki/Reporting-Visual-Studio-crashes-and-performance-issues#crashes">https://github.com/dotnet/roslyn/wiki/Reporting-Visual-Studio-crashes-and-performance-issues#crashes</a></p>

#### Timtim Bwasheze on 11/15/2019, 03:19 AM: 

Visual Studio was crashing because it cannot handle files that big(10mb) class file. Which I wish it could. I split the file in 2 but it was still crashed. Then I went on the split the file into 12 partial classes that helped, as it did not crash anymore but opening the bigger files and scrolling up and down slowed down visual studio a bit. Finally I refactored those partial classes into 13 seperate class files, since then everything seems to be ok. I think that VS was crashing because the 1 monolith file was using too much memory, causing vs to become unresponsive, eventually crashing. Visual Studio Code on the other hand can open the file no problem and also alows scrolling up and down smoothly enogh.
#### jamiehankins on 4/8/2020, 01:55 PM: 

I have a tool that generates test code. If the source file is big (&gt; 30,000 lines), then Visual Studio becomes unstable and crashes.<br><br>The next page asks for an attachment. Unfortunately, I can&#39;t include my company&#39;s file.<br><br>However, I reproed the issue by making a file like this:<br>public class TestClass<br>{<br>    public void PrintStuff()<br>    {<br>        #region Stuff To Print<br>        string item = &quot;item&quot;<br>        .ToLower()<br>        /* Insert 30,000 more .ToLower() here. */<br>        .ToLower();<br>        #endregion<br>    }<br>}<br><br>I know that this isn&#39;t a normal scenario, but this does crash with what is a normal scenario. If it doesn&#39;t crash right away, try closing or opening the region. By the way, it&#39;s not good that you don&#39;t preserve line-endings in these reports.<br>
#### Feedback Bot on 4/15/2020, 08:01 AM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

#### Feedback Bot on 5/8/2020, 00:21 AM: 

I have detected that for the last 35 days, this issue didn't have much product team activity and a very small amount of new votes or comments. Based on this, its severity, and affected area, it’s my experience that this issue is very unlikely to be fixed.
#### jamiehankins on 5/8/2020, 00:32 AM: 

<p>So you're good with an easily reproducible crashing bug in Visual Studio? </p>

#### jamiehankins on 5/18/2020, 10:57 AM: 

It&#39;s really a shame that an easily reproducible crashing bug is something that the Visual Studio team is happy to ignore.<br><br>Would it be more likely to get fixed if I figure out a way to classify it as a denial of service bug?

---
#### jamiehankins solved on 5/8/2020, 00:31 AM, 0 votes: 

<p>Do you're good with having an easily reproducible crashing bug in Visual Studio?</p>

**Comment by CyrusNajmabadi** [comment]
@jamiehankins.  In general we will prioritize our efforts based on general impact to hte community *as a whole*.  i.e. tehre are crashes/hangs/perf issues impacting thousands to millions of users.  THere are also unpleasant bugs (though less severe) that are impacting all customers.  These would take priority over a crash only affecting a very narrow slice of users.

That said, in cases like this, we would also gladly take community contributions that fix issues that are important, but not as high a priority as all the rest of hte work we have in front of us.  So if you wanted to contribute a fix here to unblock yourself, we would very likely take it (as long as it wasn't huge complex, or problematic in other areas).

Again, this doesn't mean we don't care about this issue or think that it's good that it's happening.  It's just that we have fixed amount of time to fix anything or work on anything, and i can assure you that there are much more important issues hitting many more users.  So this is simply further behind in the fix list for the time being.  The only options are to either wait until we get to this (which might take a while), or see if you can personally address this yourself through a contribution.

Thanks, and sorry we don't hve better news for you!

**Comment by jamiehankins** [comment]
@CyrusNajmabadi, thank you for your thoughtful response. I very much appreciate the process of prioritizing work, so I definitely get it.

I'm chasing my own deadlines right now, but I would love to dig in and see if there's an easy fix for this when I have a couple of hours to come up for air. I can certainly dig around myself, but is there a "getting started" document you have a pointer handy to? I'd need to build and run a debug version of the IDE.

Thanks,
Jamie

**Comment by CyrusNajmabadi** [comment]
Yup!  https://github.com/dotnet/roslyn/blob/master/docs/wiki/Contributing-Code.md and https://github.com/dotnet/roslyn/blob/master/docs/wiki/Building-Testing-and-Debugging.md.  Look at the bottom of https://github.com/dotnet/roslyn for helpful community links to where you can ask more questions and interact with team members :)

**Comment by jamiehankins** [comment]
So this particular crash is because you recursively call Microsoft.CodeAnalysis.Wrapping.ChainedExpression.Decompose(). Recursion is a dangerous thing if you can't control how deep it might go.

Our real case (not the contrived one I posted above) involves getting transactions from a database and generating code from chaining them together. That works great most of the time, but breaks down when we import a seriously big database. I can work-around this by changing our tool to not use chaining for the generated source.

Perhaps changing this to be iterative rather than recursive would solve the problem.

**Comment by CyrusNajmabadi** [comment]
> Recursion is a dangerous thing if you can't control how deep it might go

Yup, and in general we have dozens (if not more) components that recurse on the tree. Due to the generally huge cost to switch those to something else, we've effectively decided that it's out of scope to support this for very large trees *unless* it's pretty simple to handle.

If things are simple ehre, go for it :)

**Comment by CyrusNajmabadi** [comment]
Looking here, it seems very easy to make iterative.  So a PR for this would def be taken.

**Comment by jamiehankins** [comment]
Done (https://github.com/dotnet/roslyn/pull/48824).

At work, we use Azure DevOps, so this is my first Github PR. Let me know if I screwed anything up. :-)

**Comment by CyrusNajmabadi** [comment]
Fixed with https://github.com/dotnet/roslyn/pull/49096.
